In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/auto-sales-data/Auto Sales data.csv


In [2]:
# Importing necessary libraries
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler

# Creating a Linear Regression model
lm = LinearRegression()
lm

LinearRegression()

In [3]:
# Loading the dataset
df = pd.read_csv('/kaggle/input/auto-sales-data/Auto Sales data.csv')

In [4]:
df = df.drop(['ORDERDATE','POSTALCODE','CONTACTLASTNAME', 'CONTACTFIRSTNAME','ORDERNUMBER','PHONE', 'ADDRESSLINE1','POSTALCODE'],axis=1)
df.head()

QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  DAYS_SINCE_LASTORDER  \
0               30      95.70                2  2871.00                   828   
1               34      81.35                5  2765.90                   757   
2               41      94.74                2  3884.34                   703   
3               45      83.26                6  3746.70                   649   
4               36      96.66                1  3479.76                   586   

    STATUS  PRODUCTLINE  MSRP PRODUCTCODE          CUSTOMERNAME        CITY  \
0  Shipped  Motorcycles    95    S10_1678     Land of Toys Inc.         NYC   
1  Shipped  Motorcycles    95    S10_1678    Reims Collectables       Reims   
2  Shipped  Motorcycles    95    S10_1678       Lyon Souveniers       Paris   
3  Shipped  Motorcycles    95    S10_1678     Toys4GrownUps.com    Pasadena   
4  Shipped  Motorcycles    95    S10_1678  Technics Stores Inc.  Burlingame   

  COUNTRY DEALSIZE  
0     USA    Small  
1  France    Small  
2  France   Medium  
3     USA   Medium  
4     USA   Medium

In [5]:
df.nunique()

QUANTITYORDERED           58
PRICEEACH               1843
ORDERLINENUMBER           18
SALES                   2690
DAYS_SINCE_LASTORDER    1832
STATUS                     6
PRODUCTLINE                7
MSRP                      80
PRODUCTCODE              109
CUSTOMERNAME              89
CITY                      71
COUNTRY                   19
DEALSIZE                   3
dtype: int64

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2747 entries, 0 to 2746
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   QUANTITYORDERED       2747 non-null   int64  
 1   PRICEEACH             2747 non-null   float64
 2   ORDERLINENUMBER       2747 non-null   int64  
 3   SALES                 2747 non-null   float64
 4   DAYS_SINCE_LASTORDER  2747 non-null   int64  
 5   STATUS                2747 non-null   object 
 6   PRODUCTLINE           2747 non-null   object 
 7   MSRP                  2747 non-null   int64  
 8   PRODUCTCODE           2747 non-null   object 
 9   CUSTOMERNAME          2747 non-null   object 
 10  CITY                  2747 non-null   object 
 11  COUNTRY               2747 non-null   object 
 12  DEALSIZE              2747 non-null   object 
dtypes: float64(2), int64(4), object(7)
memory usage: 279.1+ KB


In [7]:
label_encoder = LabelEncoder()

df['STATUS'] = label_encoder.fit_transform(df['STATUS'])
df['PRODUCTLINE'] = label_encoder.fit_transform(df['PRODUCTLINE'])
df['PRODUCTCODE'] = label_encoder.fit_transform(df['PRODUCTCODE'])
df['CUSTOMERNAME'] = label_encoder.fit_transform(df['CUSTOMERNAME'])
df['CITY'] = label_encoder.fit_transform(df['CITY'])
df['COUNTRY'] = label_encoder.fit_transform(df['COUNTRY'])
df['DEALSIZE'] = label_encoder.fit_transform(df['DEALSIZE'])

df.head()

QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  DAYS_SINCE_LASTORDER  \
0               30      95.70                2  2871.00                   828   
1               34      81.35                5  2765.90                   757   
2               41      94.74                2  3884.34                   703   
3               45      83.26                6  3746.70                   649   
4               36      96.66                1  3479.76                   586   

   STATUS  PRODUCTLINE  MSRP  PRODUCTCODE  CUSTOMERNAME  CITY  COUNTRY  \
0       5            1    95            0            44    41       18   
1       5            1    95            0            63    55        6   
2       5            1    95            0            45    51        6   
3       5            1    95            0            82    52       18   
4       5            1    95            0            76    12       18   

   DEALSIZE  
0         2  
1         2  
2         1  
3         1  
4         1

In [8]:
# 'SALES' is the target variable
X = df.drop('SALES', axis=1)  # Features
y = df['SALES']  # Target variable

In [9]:
X.head()

QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER  DAYS_SINCE_LASTORDER  STATUS  \
0               30      95.70                2                   828       5   
1               34      81.35                5                   757       5   
2               41      94.74                2                   703       5   
3               45      83.26                6                   649       5   
4               36      96.66                1                   586       5   

   PRODUCTLINE  MSRP  PRODUCTCODE  CUSTOMERNAME  CITY  COUNTRY  DEALSIZE  
0            1    95            0            44    41       18         2  
1            1    95            0            63    55        6         2  
2            1    95            0            45    51        6         1  
3            1    95            0            82    52       18         1  
4            1    95            0            76    12       18         1

In [10]:
X.nunique()

QUANTITYORDERED           58
PRICEEACH               1843
ORDERLINENUMBER           18
DAYS_SINCE_LASTORDER    1832
STATUS                     6
PRODUCTLINE                7
MSRP                      80
PRODUCTCODE              109
CUSTOMERNAME              89
CITY                      71
COUNTRY                   19
DEALSIZE                   3
dtype: int64

In [11]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2747 entries, 0 to 2746
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   QUANTITYORDERED       2747 non-null   int64  
 1   PRICEEACH             2747 non-null   float64
 2   ORDERLINENUMBER       2747 non-null   int64  
 3   DAYS_SINCE_LASTORDER  2747 non-null   int64  
 4   STATUS                2747 non-null   int64  
 5   PRODUCTLINE           2747 non-null   int64  
 6   MSRP                  2747 non-null   int64  
 7   PRODUCTCODE           2747 non-null   int64  
 8   CUSTOMERNAME          2747 non-null   int64  
 9   CITY                  2747 non-null   int64  
 10  COUNTRY               2747 non-null   int64  
 11  DEALSIZE              2747 non-null   int64  
dtypes: float64(1), int64(11)
memory usage: 257.7 KB


In [12]:
# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [13]:
X_train.head()

QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER  DAYS_SINCE_LASTORDER  \
1559               23     145.01               10                  1585   
790                38     138.58               12                   989   
1770               40      94.71                2                  2200   
480                49      43.13                6                  1166   
2033               33      84.77               15                  2679   

      STATUS  PRODUCTLINE  MSRP  PRODUCTCODE  CUSTOMERNAME  CITY  COUNTRY  \
1559       3            3   122           61            36     6       18   
790        5            1    60           31            44    41       18   
1770       5            6    88           70            77    46       18   
480        5            6    53           19            72    15        0   
2033       5            2    72           81            82    52       18   

      DEALSIZE  
1559         1  
790          1  
1770         1  
480          2  
2033         2

In [14]:
# Standardizing of features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [15]:
# Training the model
lm.fit(X_train_scaled, y_train)

LinearRegression()

In [16]:
# Making predictions on the test set
y_pred = lm.predict(X_test_scaled)
y_pred

#y_pred_new = abs(y_pred)

array([2357.29753251, 1830.94617495, 2581.7778716 , 2308.61514123,
       1914.98977822, 4354.72041117, 1956.78838805, 6885.18895894,
       3391.06396618, 5983.08814744, 3745.2933092 , 2342.50596464,
        773.64986087, 3317.04413668, 3378.47245844, 5319.90416449,
       6330.8309736 , 2852.09245201, 6255.86347742, 2703.14687406,
       6264.29452533, 2995.18328327, 2262.7914225 , 2730.59425498,
       1748.17046659, 1547.759916  , 3354.70658871, 6123.74114427,
       2771.43468756, 4755.27899005, 3434.22118837, 6196.54278212,
       5340.49054115, 4513.2903271 , 4638.14307859, 3788.92760543,
       5041.39972907, 5532.00071266, 2189.40654653, 1539.95825969,
       1509.3001944 , 1146.93123616, 2689.74673305, 4428.71512122,
       2678.25965476, 2319.07209961, 4403.75868411, 1814.02433528,
       6893.41871285, 3440.38232024, 3407.85205571, 1638.06660477,
       2087.69188877,  986.94937966, 5636.04129558, 2658.90468031,
       6372.42222058, 5968.30174798, 1301.27268247, 1620.21433

In [17]:
# Evaluating the model
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Absolute Error: {mae}')
print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')

Mean Absolute Error: 288.76389281091616
Mean Squared Error: 163614.01426198267
R-squared: 0.9463268634133618


**Final Results**

Mean Absolute Error: 288.76389281091616

Mean Squared Error: 163614.01426198267

R-squared: 0.9463268634133618